Support `triton.language.dtype` with `torch.compile`

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121739
* __->__ #121690

Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

oulgen committed:
triton.dtype

[ghstack-poisoned]

oulgen committed:
Update on "[RFC] Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR helps towards Mamba2 execution with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Skylion007 reviewed (approved):
We should probably upstream some of these fixes to triton.

oulgen commented:
> We should probably upstream some of these fixes to triton

I'm not sure how we would upstream these changes. Maybe we can create an alias such that fp32 means float32 etc?

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -1222,6 +1222,13 @@ def add_code_part(code_part, guard, log_only=False):
         globals_for_guard_fn = {"G": builder.scope["G"]}
+        from torch.utils._triton import has_triton
+
+        if has_triton() and "triton" in pycode:
```
```suggestion
        if has_triton():
```
This will make it not break @anijain2305's C++ guards

oulgen replied:
ah we didn't need the "triton" in pycode, that was left over from my debugging, we can unconditionally add triton to the global scope

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1013,6 +1013,15 @@ def define_kernel(
     def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
+        import triton
+
+        # Horrible Hack
+        # Triton emits triton.language.float32 as triton.language.fp32
+        # so our code generation fails as fp32 is not a real thing
```
Submit a PR to triton to add an fp32 alias (or fix the repr print)?  Then add a TODO in this code linking to the PR (so we can remove the hack in the future).

oulgen replied:
Created https://github.com/openai/triton/pull/3342

jansel commented on torch/fx/graph.py:
```diff
@@ -380,11 +380,21 @@ def additional_globals(self) -> List[Tuple[str, Any]]:
+        if has_triton():
+            import triton
+            globals_[triton.__name__] = triton
+            # Horrible Hack
+            # Triton emits triton.language.float32 as triton.language.fp32
+            # so our code generation fails as fp32 is not a real thing
+            triton.language.dtype.__str__ = lambda self: dtype_to_string(self)
```
Rather than repeating this code, lets move it to a common place and call it.

jansel commented:
Tests are failing

oulgen committed:
Update on "[RFC] Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "[RFC] Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "[RFC] Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen committed:
Update on "Support `triton.language.dtype` with `torch.compile`"


Putting this PR as an RFC since I have resorted to some horrible hacks in order to make this work.
```
(Pdb) p triton.language.float32
triton.language.fp32
(Pdb) p str(triton.language.float32)
'fp32'
(Pdb) p repr(triton.language.float32)
'triton.language.fp32'
```
This means that we need to "rewrite" them for fx graph and inductor execution.

This PR allows Mamba2 to work with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge -i

houseroad commented on torch/fx/graph.py:
```diff
@@ -380,11 +380,19 @@ def additional_globals(self) -> List[Tuple[str, Any]]:
         wrapped_fns: Dict[str, None] = {}
 
+        if has_triton():
+            import triton
```
can we lift "import triton" to common file, and only import it once at the beginning?
